presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/ElasticsearchMetadata.java

```diff
@@ -114,41 +111,90 @@ public ElasticsearchTableHandle getTableHandle(ConnectorSession session, SchemaT
     @Override
     public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle table)
     {
-        ElasticsearchTableHandle handle = (ElasticsearchTableHandle) table;
-        return getTableMetadata(handle.getSchema(), handle.getIndex());
+        return getInternalTableMetadata(table).getConnectorTableMetadata();
     }
 
-    private ConnectorTableMetadata getTableMetadata(String schemaName, String tableName)
+    private InternalTableMetadata getInternalTableMetadata(ConnectorTableHandle table)
     {
-        IndexMetadata metadata = client.getIndexMetadata(tableName);
+        ElasticsearchTableHandle handle = (ElasticsearchTableHandle) table;
+        return getInternalTableMetadata(handle.getSchema(), handle.getIndex());
+    }
 
-        return new ConnectorTableMetadata(
-                new SchemaTableName(schemaName, tableName),
-                toColumnMetadata(metadata));
+    private InternalTableMetadata getInternalTableMetadata(String schemaName, String index)
+    {
+        return new InternalTableMetadata(
+                new SchemaTableName(schemaName, index),
+                client.getIndexMetadata(index));
     }
 
-    private List<ColumnMetadata> toColumnMetadata(IndexMetadata metadata)
+    private class InternalTableMetadata
```

Comment:
Move to the bottom of this file

Reply:
There's no good reason for this class to be non-static and for the logic in getColumnMetadata/getColumnHandles to be on-demand. 

Make the class a pure structure (and make it static), and initialize it with the handles and metadata. So, something like:

```java
private InternalTableMetadata makeInternalTableMetadata(SchemaTableName table, IndexMetadata metadata)
{
    // build List<ColumnMetadata> from IndexMetadata
    List<ColumnMetadata> columnMetadata = ...;
    // build Map<String, ColumnHandle> from IndexMetadata
    Map<String, ColumnHandle> columnHandles = ...;

    return new InternalTableMetadata(table, columnMetadata, columnHandles)
}
```

Reply:
Done. I added `SchemaTableName` to `InternalTableMetatadata` as you suggest but it is not used in the codebase. It is natural to be there though, and may come handy for object introspection.